kroxylicious-filters/kroxylicious-encryption/src/test/java/io/kroxylicious/filter/encryption/dek/CipherSpecTest.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.filter.encryption.dek;
+
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.security.GeneralSecurityException;
+import java.util.Arrays;
+import java.util.stream.Collectors;
+
+import javax.crypto.Cipher;
+import javax.crypto.KeyGenerator;
+import javax.crypto.SecretKey;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.EnumSource;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+class CipherSpecTest {
+
+    @Test
+    void fromPersistentIdShouldThrowIfUnknownPersistentId() {
+        assertThatThrownBy(() -> CipherSpec.fromPersistentId(123)).isExactlyInstanceOf(UnknownCipherSpecException.class);
+    }
+
+    @Test
+    void persistentIdsShouldBeUnique() {
+        assertThat(Arrays.stream(CipherSpec.values()).map(CipherSpec::persistentId).collect(Collectors.toSet()))
+                .hasSize(CipherSpec.values().length);
+    }
+
+    @ParameterizedTest
+    @EnumSource(CipherSpec.class)
+    void serializedParamsGoodForDecrypt(CipherSpec spec) throws GeneralSecurityException {
+
+        assertThat(CipherSpec.fromPersistentId(spec.persistentId())).isSameAs(spec);
+
+        var params = spec.paramSupplier().get();
+
+        // TODO Right now we're getting away with using AES keys with ChaCha20 cipher
+        // (It doesn't work the other way around)
+        // Perhaps CipherSpec needs to abstract key generation and key deserialization
+        // And the Kms should take the kind of SecretKey as a parameter
+        // and use that in how it deserializes keys when generating DekPair and/or unwrapping DEKs
+        // Or perhaps it would be more honest for the KMS to return SecretKeySpec
+        // ?
```

Comment:
I had to do some reading to figure out what this is describing.

The KMS currently returns a SecretKey. There is no guaranteed that the SecretKey from the KMS will be compatible with all ciphers.

The ChaCha20 cipher requires a Key with `format=RAW` and length 32 bytes, so it just happens to work with the AES keys we produce in the in-memory KMS.

AES on the other hand would complain if the SecretKey did not have an `algorithm == AES`. I guess it's just a sanity check that your key is intended for AES and should be of one of the supported lengths?

So to ensure we can support various Cipher/SecretKey combinations you are looking for some more flexible way to get from the bytes in the KMS to the SecretKey plugged into the Cipher? A 32-byte key from the KMS should be valid for both AES and ChaCha20 without the KMS impl having to get these details right. What if the KMS returned byte[] and we left it up to the filter to handle the key creation?